### PR TITLE
DM-14806: bring demo to standards

### PR DESCRIPTION
**Pull request by SimonKrughoff**
This also takes care of: DM-15044

### CODE REVIEW

File: tests/SConscript [hunk 1]
```diff
@@ -0,0 +1,3 @@
+# -*- python -*-
+from lsst.sconsUtils import scripts
+scripts.BasicSConscript.tests()
```
Comment (timj) [inline tests/SConscript]: Can you add `pyList=[]` argument so auto file detection is turned on? Then add a `setup.cfg` so that we can flake8 the python files.

File: tests/test_demo.py [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+# You should have received a copy of the LSST License Statement and
+# the GNU General Public License along with this program.  If not,
+# see <http://www.lsstcorp.org/LegalNotices/>.
+#
+
+import pytest
+
+import unittest
+
+from lsst.utils import tests
```
Comment (timj) [inline tests/test_demo.py]: The idiom we use is `import lsst.utils.tests` and then be explicit below.

File: tests/test_demo.py [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+
+executable_dir = 'bin'
```
Comment (timj) [inline tests/test_demo.py]: Please determine this relative to this test file so that the tests can run from outside the demo root.

File: ups/lsst_dm_stack_demo.build [hunk 4]
```diff
@@ -0,0 +1,2 @@
+@LSST BUILD@ &&
```
Comment (timj) [inline ups/lsst_dm_stack_demo.build]: Is this file needed? I don't have one in `daf_butler` and everything works.

File: ups/lsst_dm_stack_demo.table [hunk 5]
```diff
@@ -0,0 +1 @@
+setupRequired(obs_sdss)
```
Comment (timj) [inline ups/lsst_dm_stack_demo.table]: You need to add `utils` dependency (you use it in the tests) along with `python` and `sconsUtils`.

File: bin/export-results [hunk 6]
```diff
@@ -81,8 +81,8 @@ for filter in "ugriz":
                 v = srcs.get(key)
             else:
                 v = srcs.get(col)
-
+            v = np.asarray(v)
             vecs.append(v)
 
         for vals in zip(*vecs):
-            print(' '.join([str(el) for el in vals]))
+            print(' '.join(['{0:.12g}'.format(el) if issubclass(el.dtype.type, np.floating) else str(el) for el in vals]))
```
Comment (timj) [inline bin/export-results]: Might be worth adding a comment before this line

File: SConstruct [hunk 7]
```diff
@@ -0,0 +1,3 @@
+# -*- python -*-
+from lsst.sconsUtils import scripts
+scripts.BasicSConstruct("lsst_dm_stack_demo")
```
Comment (timj) [inline SConstruct]: Add `disableCc` since we don't need a C++ compiler.

File: tests/SConscript (hunk 1 above)
Comment (SimonKrughoff) [reply tests/SConscript]: Done, I think.

File: tests/test_demo.py (hunk 2 above)
Comment (SimonKrughoff) [reply tests/test_demo.py]: Done.

File: ups/lsst_dm_stack_demo.build (hunk 4 above)
Comment (SimonKrughoff) [reply ups/lsst_dm_stack_demo.build]: Cargo culted from ci_lsst.

File: tests/test_demo.py (hunk 3 above)
Comment (SimonKrughoff) [reply tests/test_demo.py]: You mean determine the absolute path, right?

Comment (timj) [reply tests/test_demo.py]: I mean:
```python
executable_dir = os.path.join(os.path.dirname(__file__), os.path.pardir, "bin")
```
You might need to include `os.path.abspath()` as well.

File: ups/lsst_dm_stack_demo.build (hunk 4 above)
Comment (PaulPrice) [reply ups/lsst_dm_stack_demo.build]: Drop it. We don't use the build files any more.

File: .travis.yaml [hunk 8]
```diff
@@ -0,0 +1,16 @@
+sudo: false
+language: python
+matrix:
+  include:
+    - python: '2.7'
```
Comment (timj) [inline .travis.yaml]: Remove the 2.7 and 3.5 stuff.
```
sudo: false
language: python
matrix:
  include:
    - python: '3.6'
      install:
      - pip install flake8
      script: flake8
```
is sufficient.

File: .travis.yaml [hunk 9]
```diff
@@ -0,0 +1,16 @@
...
+      - pip install flake8
+      script: flake8
+    - python: '3.5'
+      install:
+      - pip install flake8
+      script: flake8      
+    - python: '3.6'
+      install:
+      - pip install flake8
+      script: flake8
```
Comment (timj) [inline .travis.yaml]: missing newline.

File: setup.cfg [hunk 10]
```diff
@@ -0,0 +1,10 @@
+[flake8]
+max-line-length = 110
+ignore = E133, E226, E228, N802, N803
+exclude = andConfig.py, processCcd.py
+
+[tool:pytest]
+addopts = --flake8
+flake8-ignore = E133 E226 E228 N802 N803 N806
+    processCcd.py ALL
+    andConfig.py ALL
```
Comment (timj) [inline setup.cfg]: New line needed.

File: tests/test_demo.py [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+# see <http://www.lsstcorp.org/LegalNotices/>.
+#
+
+import unittest
+import os
+
+import lsst.utils.tests
+from lsst.utils import getPackageDir
+
+package_root = getPackageDir('obs_test')
```
Comment (timj) [inline tests/test_demo.py]: It's open to debate whether it's best to use `getPackageDir` or use this test file location as the reference. I tend to prefer the latter but I'm fine with the former if that's the way you want to go.

File: ups/lsst_dm_stack_demo.build [hunk 12]
```diff
@@ -0,0 +1 @@
+build_lsst @PRODUCT@ @VERSION@ @REPOVERSION@
```
Comment (timj) [inline ups/lsst_dm_stack_demo.build]: Remove this file.

File: tests/test_demo.py [hunk 13]
```diff
@@ -0,0 +1,50 @@
...
+                              args=["--small"],
+                              msg="Running demo failed")
+        self.assertExecutable("compare",
+                              root_dir=executable_dir,
+                              args=['detected-sources_small.txt'],
+                              msg="Compare failed")
+
+
+if __name__ == "__main__":
+    lsst.utils.tests.init()
```
Comment (timj) [inline tests/test_demo.py]: This is not necessary. We are not doing any tests that leak files or memory (and it's not run by pytest because there is no `setup_module` equivalent.

File: tests/test_demo.py [hunk 14]
```diff
@@ -0,0 +1,50 @@
...
+executable_dir = os.path.join(package_root, 'bin')
+
+
+class DemoTestCase(lsst.utils.tests.ExecutablesTestCase):
+    """Test the demo scrpts for executablility."""
+    def testDemo(self):
+        """Test demo"""
+        self.assertExecutable("demo.sh",
+                              root_dir=executable_dir,
+                              args=["--small"],
```
Comment (timj) [inline tests/test_demo.py]: Is it okay to use `--small` here? How do we ever test with large?

File: setup.cfg [hunk 15]
```diff
@@ -0,0 +1,10 @@
+[flake8]
+max-line-length = 110
+ignore = E133, E226, E228, N802, N803
```
Comment (timj) [inline setup.cfg]: Add N806.

**Comment by timj** [comment]
Probably should update the REAME as well.

File: ups/lsst_dm_stack_demo.table (hunk 5 above)
Comment (SimonKrughoff) [reply ups/lsst_dm_stack_demo.table]: I think I did this.

File: tests/test_demo.py (hunk 14 above)
Comment (SimonKrughoff) [reply tests/test_demo.py]: I'm going to leave as is for now as I don't know how this will be dealt with down the road.

File: .travis.yaml (hunk 8 above)
Comment (SimonKrughoff) [reply .travis.yaml]: Done

File: .travis.yaml (hunk 9 above)
Comment (SimonKrughoff) [reply .travis.yaml]: Done.

File: setup.cfg (hunk 10 above)
Comment (SimonKrughoff) [reply setup.cfg]: 👍

File: tests/test_demo.py (hunk 11 above)
Comment (SimonKrughoff) [reply tests/test_demo.py]: I decided to leave it the way it is.  If you hate it, let me know.

File: ups/lsst_dm_stack_demo.build (hunk 12 above)
Comment (SimonKrughoff) [reply ups/lsst_dm_stack_demo.build]: Done.

File: setup.cfg (hunk 15 above)
Comment (SimonKrughoff) [reply setup.cfg]: Done.

File: tests/test_demo.py (hunk 11 above)
Comment (timj) [reply tests/test_demo.py]: It had better be asking for `lsst_dm_stack_demo` package location though 😉  -- although as you've written it you are making my case for using `__file__`.

**Comment by timj** [comment]
@SimonKrughoff I rebased this branch and did a test build on Jenkins. It failed the astrometry check:
```
1112 sources in camcol : 4
1112 Sources in reference visit : 4192
1264 sources in camcol :  4
Visit : 6377 750 matches found
Median value of the astrometric scatter - all magnitudes: 169.760284819 mas
Astrometric scatter (median) - mag < 19.5 : 102.698862873 mas
Median astrometric scatter 102.7 mas is larger than reference : 100.0 mas 
```

See https://ci.lsst.codes/blue/organizations/jenkins/stack-os-matrix/detail/stack-os-matrix/28217/pipeline/46

**Comment by timj** [comment]
It also looks like `check_astrometry` tries to make a plot and this fails because `$DISPLAY` is not set on Jenkins.

**Comment by RobertLuptonTheGood** [comment]
Plots should all be controlled by lsstDebug (for now), so this sounds like an easy fix.

**Comment by timj** [comment]
The plots aren't protected by anything. It always seems to make the PNG. The fix is easy in that we just need to run with a command line option `--no-plot` or something. More important is that the check_astrometry routine doesn't return any information on pass/fail so the script exit status can't be reported to the test harness.

**Comment by wmwv** [comment]
* The key issue is that the default matplotlib on the Jenkins nodes assumes a backend that the default Jenkins nodes don't actually have.

Minor: The `import matplotlib.pylab as plt` line should be moved from a module-level import to be within `def plotAstrometry` function.
B) `main` could be updated to not plot by default.  We're not actually looking at or saving the plot.  
```
def main(repo, runs, fields, ref, ref_field, camcol, filter, plot=False):
    """Main executable.

    """

    struct = loadAndMatchData(repo, runs, fields, ref, ref_field, camcol, filter)
    mag = struct.mag
    dist = struct.dist
    match = struct.match
    checkAstrometry(mag, dist, match)
    if plot:
        plotAstrometry(mag, dist, match)
```

**Comment by wmwv** [comment]
> Plots should all be controlled by lsstDebug (for now), so this sounds like an easy fix.

Can you remind me where this policy is described and if there is a snippet of the recommended wrapper code to implement this?

I agree with using `lsstDebug` for plotting in all of the science pipelines algorithmic code, but it's less clear to me that things like this demo that are trying demonstrate the full capabilities and meant to be run both in test environments and by an end user should require using `lsstDebug` to set up a plot.

**Comment by RobertLuptonTheGood** [comment]
I was lazy and didn't read the commit -- I expected that useful plots would get added to the Task code (using lsstDebug) and enabled by a command line option, rather than being directly added to the demo framework.

**Comment by timj** [comment]
@wmwv I think your proposal agrees with mine and looks fine. We do need to set exit status if the astrometry test fails, although I'm not sure if anyone has noticed that this test is failing and what the right answer really is (and whether it matters that @SimonKrughoff changed the filter for the astrometry test to make it work with `--small`).

**Comment by timj** [comment]
I've made some tweaks and it now runs fine on Jenkins and fails because 102 > 100. https://ci.lsst.codes/blue/organizations/jenkins/stack-os-matrix/detail/stack-os-matrix/28221/tests

It's for someone else to decide that 102 mas is fine.

**Comment by timj** [comment]
This is indeed the first time we've set exit status and would notice a problem. Digging deeper the difference is the change in filter that @SimonKrughoff had to add to the check program to make it work with `--small`. With the `r` filter:

```
CameraMapper INFO: Loading exposure registry from /Users/timj/work/lsstsw3/build/lsst_dm_stack_demo/input/registry.sqlite3
777 sources in camcol : 4
777 Sources in reference visit : 4192
892 sources in camcol :  4
Visit : 6377 576 matches found
Median value of the astrometric scatter - all magnitudes: 103.87098207092725 mas
Astrometric scatter (median) - mag < 19.5 : 85.26176053780722 mas
Ok.
```
and with the `i` filter currently being used:
```
CameraMapper INFO: Loading exposure registry from /Users/timj/work/lsstsw3/build/lsst_dm_stack_demo/input/registry.sqlite3
1112 sources in camcol : 4
1112 Sources in reference visit : 4192
1264 sources in camcol :  4
Visit : 6377 750 matches found
Median value of the astrometric scatter - all magnitudes: 169.76028481841527 mas
bin.src/check_astrometry.py:226: RuntimeWarning: invalid value encountered in less
  idxs = np.where(np.asarray(mag) < good_mag_limit)
Astrometric scatter (median) - mag < 19.5 : 102.6988628735584 mas
Median astrometric scatter 102.7 mas is larger than reference : 100.0 mas 
```
I'll force this to 105mas for this filter so it will pass.

**Comment by timj** [comment]
With these changes the demo runs on Jenkins as a standard eups package and passes tests. It now doesn't work the old way (it could do with some tweaking to the demo script) but it's not clear we need it to since we will be removing the special case running of demo in Jenkins.

File: README [hunk 16]
```diff
@@ -14,7 +14,7 @@ Usage
 To run the example, do::
 
 	$ source $LSST_HOME/loadLSST.sh
-	$ setup obs_sdss
+	$ setup -r .
```
Comment (timj) [inline README]: We need to decide whether the instructions then say "run scons" or whether they stay as they are and me make `demo.sh` look in `../bin.src/` for the Python code.
